### ISSUE TITLE
Unexpected behavior of Dependency Injection from components

### ISSUE THREAD

**Issue by Xerphis** [phalcon/cphalcon#13527]
Hi.

Package with project: [phalcon-di-problem.zip](https://github.com/phalcon/cphalcon/files/2481896/phalcon-di-problem.zip)

database query
```sql
CREATE SCHEMA `test_schema`;
USE `test_schema`;
DROP TABLE IF EXISTS `test_table`;
CREATE TABLE IF NOT EXISTS `test_table` (
  `id` mediumint(3) NOT NULL AUTO_INCREMENT,
  `content` varchar(65) NOT NULL,
  PRIMARY KEY (`id`)
) ENGINE=MyISAM AUTO_INCREMENT=2 DEFAULT CHARSET=latin1;
INSERT INTO `test_table` (`id`, `content`) VALUES (1, 'test');
```

public/index.php
```php
function dump($foo) {
	echo (new \Phalcon\Debug\Dump())->variable($foo);
}

// Dependency Inject
$di = new \Phalcon\Di\FactoryDefault();

// Debug
(new \Phalcon\Debug())->listen();

// Loader
$di->setShared('loader', function () {
	$loader = new \Phalcon\Loader();
	$loader->registerNamespaces([
			'Controllers' => '../app/controllers',
			'Models' => '../app/models',
			'Libraries' => '../app/libraries'
	]);
	$loader->register();
	return $loader;
});

// Router service
$di->setShared('router', function() {
	$router = new \Phalcon\Mvc\Router();
	return $router;
});

// Dispatcher service
$di->setShared('dispatcher', function () {
	$dispatcher = new \Phalcon\Mvc\Dispatcher();
	$dispatcher->setDefaultNamespace('Controllers');
	return $dispatcher;
});

// View service
$di->setShared('view', function () {
	$view = new \Phalcon\Mvc\View();
	return $view;
});

// Database service
$di->setShared('db', function () {
	return new \Phalcon\Db\Adapter\Pdo\Mysql([
		'host' => 'localhost',
		'port' => 3306,
		'username' => 'root',
		'password' => '',
		'dbname' => 'test_schema',
		'charset' => 'utf8'
	]);
});

$di->setShared('metaData', function () {
	return new \Phalcon\Mvc\Model\Metadata\Files([
		'metaDataDir' => '../app/',
	]);
});

// Auth library service
$di->setShared('auth', function () {
	$auth = new \Libraries\Auth();
	//$auth->autologin();
	return $auth;
});

// Format library service
$di->setShared('format', function () {
	$format = new \Libraries\Format();
	return $format;
});

// App run
$app = new \Phalcon\Mvc\Application();
$app->setDI($di);
echo $app->handle()->getContent();
```

controllers/IndexController.php
```php
namespace Controllers;

class IndexController extends \Phalcon\Mvc\Controller {
	public function indexAction() {
		dump('IndexController indexAction');
		$x = \Models\Test::findFirst();
		dump($x);
	}
}
```

models/Test.php
```php
namespace Models;

class Test extends \Phalcon\Mvc\Model {

	public $format;
	protected $id;
	protected $content;

	public function initialize() {
		$this->setSchema("test_schema");
		$this->setSource("test_table");
	}

	public function afterFetch() {
		dump('Test afterFetch');
		$this->format = new \Libraries\Format();
	}
}
```

libraries\Auth.php
```php
namespace Libraries;

class Auth extends \Phalcon\Mvc\User\Component {

	public $user;

	public function __construct() {
		dump('Auth construct');
		$this->autologin();
	}

	public function autologin() {
		dump('Auth autologin');
		// if is "logged", get user data...
		if (true) {
			$this->user = \Models\Test::findFirst(1);
		}
	}
}
```

libraries\Format.php
```php
namespace Libraries;

class Format extends \Phalcon\Mvc\User\Component {

	public $logged;

	public function __construct() {
		dump('Format construct');
		$auth = $this->getDI()->getShared('auth');
		dump($auth);
	}
}
```

Ok... It's a only example. I have a "auth" library. In this library, i want recieve user data (if we are logged). But, I use formatting for every model after the fetch (or after save).

The problem occurs when in the "auth" library, I want execute select query. Then, after the fetch model, the "format" object is created. In the "format" constructor, referring again to service  "auth", which I would like to call method or poperty (for example; to check, if the user is logged in - and on this basis to format the data accordingly).

Unfortunately, at this moment there is a loop - every time when in the constructor "auth" I call the model (which again refers to "auth" in the object "format"). The instance of the "auth" service is shared - constructor should be called only once. In this case, the service is created every time. The same situation is when I register the service "auth" and i call to the "autologin" method before return instance. What is wrong?

I expected that in the "format" object I would receive the same instance of the "auth" service.

### Details
* Phalcon version: 3.4.1
* PHP Version: PHP 7.2.4 (cli)
* Operating System: Windows 10 Home 64 bit
* Server: Apache 2.4.33

**Comment by Xerphis** [comment]
Hm. Ok, i set "\Phalcon\Di::setDefault($di);" before create Application object. No changes.

In fact, each model has its own format class, so I do not want to register it as a service.

I set format as service (btw, I changed the property name to "_formater", because the "format" property was called the same as the service and could have caused this complication).
models/test.php
```php
	protected $_formater;
	public function afterFetch() {
		$this->_formater = $this->getDI()->get('format');
	}
```
Shared or not shared... No changes.

I would like to add that if I do not execute a query (id do not lead to a call afterFetch model) in the auth class constructor, only for example in the controller, everything works fine: in the "format" service, call to DI "$this->getDI()->getShared('auth')" returns the same instance of Auth class in each separate format.
